pkg/manager/config/manager.go

```diff
@@ -96,103 +77,88 @@ func (srv *ConfigManager) Init(ctx context.Context, addrs []string, cfg config.A
 	return nil
 }
 
-func (e *ConfigManager) watch(ctx context.Context, ns, key string, f func(*zap.Logger, *clientv3.Event)) {
-	wkey := path.Join(e.basePath, ns, key)
-	logger := e.logger.With(zap.String("component", wkey))
+func (e *ConfigManager) watch(ctx context.Context, ns, key string, f func(*zap.Logger, mvccpb.Event)) {
+	wkey := []byte(path.Join(e.basePath, ns, key))
+	logger := e.logger.With(zap.String("component", string(wkey)))
 	retryInterval := 5 * time.Second
 	e.wg.Run(func() {
-		var prevKV *mvccpb.KeyValue
-
-		ticker := time.NewTicker(e.watchInterval)
-		defer ticker.Stop()
-
-		wch := e.etcdClient.Watch(ctx, wkey)
+		wch := e.kv.NewWatchStream()
+		defer wch.Close()
 		for {
+			var err error
 			select {
 			case <-ctx.Done():
 				return
-			case <-ticker.C:
-				resp, err := e.kv.Get(ctx, wkey)
-				if err != nil {
-					logger.Warn("failed to poll", zap.Error(err))
-					break
-				}
-				// len == 0 may mean there is no value set yet, do not warn about that
-				if len(resp.Kvs) > 1 {
-					logger.Warn("failed to poll", zap.Error(ErrNoOrMultiResults))
-					break
-				} else if len(resp.Kvs) == 1 {
-					f(logger, &clientv3.Event{
-						Type:   mvccpb.PUT,
-						Kv:     resp.Kvs[0],
-						PrevKv: prevKV,
-					})
-					prevKV = resp.Kvs[0]
-				}
-			case res := <-wch:
-				if res.Canceled {
-					// don't wait for more than the polling interval
-					if k := retryInterval * 2; k < e.watchInterval {
-						retryInterval = k
-					}
-					logger.Warn("failed to watch, will try again later", zap.Error(res.Err()), zap.Duration("sleep", retryInterval))
-					time.Sleep(retryInterval)
-					wch = e.etcdClient.Watch(ctx, wkey, clientv3.WithCreatedNotify())
-					break
-				}
-
+			default:
+				_, err = wch.Watch(mvcc.AutoWatchID, wkey, getPrefix(wkey), wch.Rev()-1)
+			}
+			if err == nil {
+				break
+			}
+			if k := retryInterval * 2; k < e.watchInterval {
+				retryInterval = k
+			}
+			logger.Warn("failed to watch, will try again later", zap.Duration("sleep", retryInterval))
+			time.Sleep(retryInterval)
```

Comment:
```go
select {
    case <-ctx.Done():
        return
    case <-time.After(retryInterval):
}
```

Reply:
Not really what I want. It will wait for `time.After(retryInterval)` before trying to watch the channel. My logic is that:

1. if context cancelled, exit
2. if not, try to watch, and sleep on failure

Reply:
What if the context is canceled during sleeping? It won't quit in time.

Reply:
Done.